src/Algebra/Properties/Semigroup/Reasoning.agda

```diff
@@ -0,0 +1,139 @@
+  [u∙v]∙[w∙x]≈u∙[[v∙w]]∙x] = begin
+    (u ∙ v) ∙ (w ∙ x) ≈⟨ assoc u v (w ∙ x) ⟩
+    u ∙ (v ∙ (w ∙ x)) ≈⟨ ∙-congˡ (sym (assoc v w x)) ⟩
+    u ∙ ((v ∙ w) ∙ x) ∎
+
+  u∙[[v∙w]∙x]≈[[u∙v]∙w]∙x : u ∙ ((v ∙ w) ∙ x) ≈ ((u ∙ v) ∙ w) ∙ x
+  u∙[[v∙w]∙x]≈[[u∙v]∙w]∙x = sym (trans (∙-congʳ (assoc u v w)) (assoc u (v ∙ w) x))
+
+  u∙[v∙w]]∙x]≈[u∙v]∙[w∙x] : u ∙ ((v ∙ w) ∙ x) ≈ (u ∙ v) ∙ (w ∙ x)
+  u∙[v∙w]]∙x]≈[u∙v]∙[w∙x] = begin
+    u ∙ ((v ∙ w) ∙ x) ≈⟨ ∙-congˡ (assoc v w x) ⟩
+    u ∙ (v ∙ (w ∙ x)) ≈⟨ assoc u v (w ∙ x) ⟨
+    (u ∙ v) ∙ (w ∙ x) ∎
+
+  u∙[v∙[w∙x]]≈[[u∙v]∙w]∙x : u ∙ (v ∙ (w ∙ x)) ≈ ((u ∙ v) ∙ w) ∙ x
+  u∙[v∙[w∙x]]≈[[u∙v]∙w]∙x = sym (trans (assoc (u ∙ v) w x) (assoc u v (w ∙ x)))
+
+  u∙[v∙[w∙x]]≈[u∙[v∙w]]∙x : u ∙ (v ∙ (w ∙ x)) ≈ (u ∙ (v ∙ w)) ∙ x
+  u∙[v∙[w∙x]]≈[u∙[v∙w]]∙x = sym (trans (assoc u (v ∙ w) x) (∙-congˡ (assoc v w x)))
+
+open Assoc4 public
+
+module Pulls (uv≡w : u ∙ v ≈ w) where
+    pullʳ : ∀ {x} → (x ∙ u) ∙ v ≈ x ∙ w
+    pullʳ {x = x} = begin
+        (x ∙ u) ∙ v ≈⟨ assoc x u v ⟩
+        x ∙ (u ∙ v) ≈⟨ ∙-congˡ uv≡w ⟩
+        x ∙ w       ∎
+
+    pullˡ : ∀ {x} → u ∙ (v ∙ x) ≈ w ∙ x
+    pullˡ {x = x} = begin
+        u ∙ (v ∙ x) ≈⟨ assoc u v x ⟨
+        (u ∙ v) ∙ x ≈⟨ ∙-congʳ uv≡w ⟩
+        w ∙ x       ∎
+
+    pull-first : ∀ {x y} → u ∙ ((v ∙ x) ∙ y) ≈ w ∙ (x ∙ y)
+    pull-first {x = x} {y = y} = begin
+      u ∙ ((v ∙ x) ∙ y) ≈⟨ ∙-congˡ (assoc v x y) ⟩
+      u ∙ (v ∙ (x ∙ y)) ≈⟨ pullˡ ⟩
+      w ∙ (x ∙ y)       ∎
+
+    pull-center : ∀ {x y} → x ∙ (u ∙ (v ∙ y)) ≈ x ∙ (w ∙ y)
+    pull-center {x = x} {y = y} = ∙-congˡ (pullˡ)
+
+    -- could be called pull₃ʳ
+    pull-last : ∀ {x y} → (x ∙ (y ∙ u)) ∙ v ≈ x ∙ (y ∙ w)
+    pull-last {x = x} {y = y} = begin
+      (x ∙ (y ∙ u)) ∙ v ≈⟨ assoc x (y ∙ u) v ⟩
+      x ∙ ((y ∙ u) ∙ v) ≈⟨ ∙-congˡ (pullʳ {x = y}) ⟩
+      x ∙ (y ∙ w)       ∎
+
+open Pulls public
+
+module Pushes (uv≡w : u ∙ v ≈ w) where
+    pushʳ : x ∙ w ≈ (x ∙ u) ∙ v
+    pushʳ = sym (pullʳ uv≡w)
+
+    pushˡ : w ∙ x ≈ u ∙ (v ∙ x)
+    pushˡ = sym (pullˡ uv≡w)
+
+open Pushes public
+
+module Center (eq : u ∙ v ≈ w) where
+  center : (x ∙ u) ∙ (v ∙ y) ≈ x ∙ (w ∙ y)
+  center = pullʳ (pullˡ eq)
+
+  center⁻¹ : x ∙ y ≈ z →  u ∙ ((v ∙ x) ∙ y) ≈ w ∙ z
+  center⁻¹ {x = x} {y = y} {z = z} xy≈z = begin
+    u ∙ ((v ∙ x) ∙ y) ≈⟨ ∙-congˡ (pullʳ xy≈z) ⟩
+    u ∙ (v ∙ z)       ≈⟨ pullˡ eq ⟩
+    w ∙ z             ∎
+
+  push-center : x ∙ (w ∙ y) ≈ x ∙ (u ∙ (v ∙ y))
+  push-center = sym (pull-center eq)
+
+open Center public
+
+module Extends {u v w x : Carrier} (s : u ∙ v ≈ w ∙ x) where
+  extendˡ : (y ∙ u) ∙ v ≈ (y ∙ w) ∙ x
+  extendˡ {y = y} = begin
+    (y ∙ u) ∙ v ≈⟨ pullʳ s ⟩
+    y ∙ (w ∙ x) ≈⟨ assoc y w x ⟨
+    (y ∙ w) ∙ x ∎
+
+  extendʳ : u ∙ (v ∙ y) ≈ w ∙ (x ∙ y)
+  extendʳ {y = y} = begin
+    u ∙ (v ∙ y) ≈⟨ pullˡ s ⟩
+    (w ∙ x) ∙ y ≈⟨ assoc w x y ⟩
+    w ∙ (x ∙ y) ∎
+
+  extend² : ∀ y z → (y ∙ u) ∙ (v ∙ z) ≈ (y ∙ w) ∙ (x ∙ z)
+  extend² y z = begin
+    (y ∙ u) ∙ (v ∙ z) ≈⟨ pullʳ (extendʳ {y = z}) ⟩
```

Comment:
This step is perhaps where the use of *implicit* `y` in `extendʳ` is a mistake: the need to make it explicit is perhaps/probably a code-smell, part of the general problem that in any *concrete* instantiation of `Semigroup`, the typechecker will not (necessarily) be able to infer arguments to any sub-term `y ∙ u`, so the defensive programming style woul dbe to make the argument explicit, and if necessary, use `_` at call-sites. Cf. type of `assoc` itself...

Reply:
This is probably true; I guess we can get away with it in the categorical case because the types will end up uniquely identify things?